utils/s2n_mem.c

```diff
@@ -25,32 +25,119 @@
 #include "utils/s2n_safety.h"
 
 static long page_size = 4096;
-static int use_mlock = 1;
+static bool initialized = false;
 
-int s2n_mem_init(void)
+static int s2n_mem_init_impl(void);
+static int s2n_mem_cleanup_impl(void);
+static int s2n_mem_free_no_mlock_impl(void *ptr, uint32_t size);
+static int s2n_mem_free_mlock_impl(void *ptr, uint32_t size);
+static int s2n_mem_malloc_no_mlock_impl(void **ptr, uint32_t requested, uint32_t *allocated);
+static int s2n_mem_malloc_mlock_impl(void **ptr, uint32_t requested, uint32_t *allocated);
+
+static s2n_mem_init_callback s2n_mem_init_cb = s2n_mem_init_impl;
+static s2n_mem_cleanup_callback s2n_mem_cleanup_cb = s2n_mem_cleanup_impl;
+static s2n_mem_malloc_callback s2n_mem_malloc_cb = s2n_mem_malloc_mlock_impl;
+static s2n_mem_free_callback s2n_mem_free_cb = s2n_mem_free_mlock_impl;
+
+static int s2n_mem_init_impl(void)
 {
     GUARD(page_size = sysconf(_SC_PAGESIZE));
     if (getenv("S2N_DONT_MLOCK")) {
-        use_mlock = 0;
+        s2n_mem_malloc_cb = s2n_mem_malloc_no_mlock_impl;
+        s2n_mem_free_cb = s2n_mem_free_no_mlock_impl;
     }
-
-    return 0;
+    return S2N_SUCCESS;
 }
 
-int s2n_mem_cleanup(void)
+static int s2n_mem_cleanup_impl(void)
 {
     page_size = 4096;
-    use_mlock = 1;
-    return 0;
+    s2n_mem_malloc_cb = s2n_mem_malloc_no_mlock_impl;
+    s2n_mem_free_cb = s2n_mem_free_no_mlock_impl;
+    return S2N_SUCCESS;
+}
+
+static int s2n_mem_free_mlock_impl(void *ptr, uint32_t size)
+{
+    int munlock_rc = munlock(ptr, size);
+    free(ptr);
+    GUARD(munlock_rc);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_mem_free_no_mlock_impl(void *ptr, uint32_t size)
+{
+    free(ptr);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_mem_malloc_mlock_impl(void **ptr, uint32_t requested, uint32_t *allocated)
+{
+    notnull_check(ptr);
+
+    /* Page aligned allocation required for mlock */
+    uint32_t allocate;
+
+    GUARD(s2n_align_to(requested, page_size, &allocate));
+
+    *ptr = NULL;
+    S2N_ERROR_IF(posix_memalign(ptr, page_size, allocate) != 0, S2N_ERR_ALLOC);
+    *allocated = allocate;
+
+#ifdef MADV_DONTDUMP
+    if (madvise(*ptr, *allocated, MADV_DONTDUMP) != 0) {
+        GUARD(s2n_free_impl(*ptr, *allocated));
+        S2N_ERROR(S2N_ERR_MADVISE);
+    }
+#endif
+
+    if (mlock(*ptr, *allocated) != 0) {
+        GUARD(s2n_mem_free_mlock_impl(*ptr, *allocated));
```

Comment:
Should this be `s2n_mem_free_mlock_impl` or `s2n_free_impl`? If mlocked failed, what does it mean to `munlock` that memory?

Reply:
s2n_mem_free_mlock_impl doesn't block on munlock. I figured it might be best just in case mlock would partially lock some pages before hitting a ulimit or something, but now I see that mlock() has to be all-or-nothing. I'll change it.